### ISSUE TITLE
AG Grid JS error when running in native window

### ISSUE THREAD

**Issue by juledwar** [zauberzeug/nicegui#3935]
### Description

This only happens when running in a native window. I have an ag-grid on a page, and when loading the page I get this error in the console output:
```
js: TypeError: Failed to resolve module specifier "ag-grid-community". Relative references must start with either "/", "./", or "../".
```

This halts further set up of the page (anything constructed after the error does not appear).

I'm also seeing
```
Registered new object after initialization, existing clients won't be notified!
```
which seems to be a Qt WebEngine error, not sure if related to this.

Can be reproduced with this code (I just copy pasted what I am doing, you can probably repro without the extra classes and settings):
```py
from nicegui import ui

ui.aggrid(
    {
        'columnDefs': [
            {
                'field': 'value',
                'type': 'numericColumn',
                'width': 100,
            },
            {'field': 'stat'},
        ],
        'defaultColDef': {
            'flex': 1,
            'enableCellChangeFlash': True,
        },
        'rowData': [],
        'headerHeight': 0,
    }
).classes('min-w-fit')

ui.run(native=True, show=False)
```

**Comment by juledwar** [comment]
nicegui==2.5.0
PyQt5==5.15.11
PyQt5-Qt5==5.15.15
PyQt5_sip==12.15.0
PyQtWebEngine==5.15.7
PyQtWebEngine-Qt5==5.15.15
QtPy==2.4.1

**Comment by falkoschindler** [comment]
Strange. I can't reproduce the problem on macOS. What OS are you using? Can anyone else reproduce the error?

**Comment by juledwar** [comment]
Ubuntu 24.04, under KDE using the QT view (not Gnome).

**Comment by mbclarkengineering** [comment]
I can reproduce on rocky linux 8.6, on firefox browser. Same issue

**Comment by evnchn** [comment]
Hey. 

Your browser may be old, or otherwise it doesn't support the ES module importmap. 

Check https://github.com/zauberzeug/nicegui/pull/4625#issuecomment-2812752236

Remind me to install Pale Moon on my laptop I brought to travel and do some testing. You can do that too. Basilisk is also another Firefox fork like Pale Moon that doesn't support ES modules (at least not completely)

**Comment by evnchn** [comment]
Reminder to self: IIRC "shim mode" of es-module-shims seem to be more compatible against dynamic imports. It would require adding `-shim` suffix to link `modulepreload`, script type `importmap` and script type `module`. 

If it means a working AG Grid for native window and legacy browsers, it may be worth switching (last time my conclusion was, shim mode does not help the `ui.markdown` nor `ui.mermaid`, but didn't tabulate my findings and memory's weak...)

**Comment by alexc155** [comment]
I have the same problem on an old Monterey 12.7.6 Mac running Safari 17.6

It seems the version of Webkit is probably too old to support ES Modules


There is a workaround however.

* Find the `aggrid.js` file in the nicegui package (mine is `.venv/lib/python3.13/site-packages/nicegui/elements/aggrid.js`)
* Remove the top line (`import "ag-grid-community";`)
* Add the contents of `ag-grid-community.js` ( `.venv/lib/python3.13/site-packages/nicegui/elements/lib/aggrid/ag-grid-community.js`) to the top of `aggrid.js`

:-)

**Bonus marks** - I also found that the native viewer is not zoomable using the usual `⌘ +` / `⌘ -` 
(even with `app.native.window_args["zoomable"] = True`)

But if you add a couple of buttons to your UI you can control the zoom:

```py
def __zoom(adjustment):
        ui.run_javascript(f"""
                            window.document.body.style["zoom"] = 
                              parseFloat(window.getComputedStyle(document.body).zoom) + parseFloat({adjustment})
                          """)

ui.add_css("""body { zoom: 1.0; }""")

ui.button("-", on_click=lambda: __zoom('-0.1'))
ui.button("+", on_click=lambda: __zoom('0.1'))
```

:-D

**Comment by falkoschindler** [comment]
@alexc155 It would be interesting to see if the problem remains on the [3.0 branch](https://github.com/zauberzeug/nicegui/tree/3.0). We reworked the node package management quite a bit and maybe the import works better now.

**Comment by alexc155** [comment]
No. The same type of error happens in https://github.com/zauberzeug/nicegui/blob/3.0/nicegui/elements/aggrid/aggrid.js

```
TypeError: Module specifier, 'nicegui-aggrid' does not start with "/", "./", or "../". Referenced from aggrid.js
```

It doesn't like importing ES Modules and it looks like the 3.0 branch is still doing that:

```js
// https://github.com/zauberzeug/nicegui/blob/3.0/nicegui/elements/aggrid/aggrid.js
import AgGrid from "nicegui-aggrid";  // <-- First line
```

**Comment by falkoschindler** [comment]
Hm, this reminds me of #5020 where removing the `nomodule` attribute in index.html helped. Somehow old webkits tend mess up this switch.

**Comment by alexc155** [comment]
Yes! Removing `nomodule` from index.html also fixes the problem. I _think_ that the version of Webkit supplied with Monterey is not the same as the version of Webkit supplied with Safari. When the code is running in normal mode, Webkit from Safari is used. When the code is running in native mode, Webkit from Monterey is used and it's older and doesn't support ES Modules / doesn't respect the `nomodule` attribute

Mac Monterey Version:

```
$ defaults read /System/Library/Frameworks/WebKit.framework/Versions/A/Resources/version.plist CFBundleVersion
17613.3.9.1.16
```

Safari Version:

```
$ defaults read /Applications/Safari.app/Contents/Info.plist CFBundleVersion
17618.3.11.11.7
```

**Comment by falkoschindler** [comment]
@alexc155 Thanks for the update! Then I assume adding the following line to the application code works as well:
```py
from nicegui import __version__, ui

ui.add_head_html(f'<script src="/_nicegui/{__version__}/static/es-module-shims.js"></script>', shared=True)
```

**Comment by alexc155** [comment]
Possibly, but even better if you sniffed the capabilities of the browser instance:

```html
    <script>
      // nomodule attribute in <script> is not accurate in MacOS Monterey - see https://github.com/zauberzeug/nicegui/issues/3935
      var nomodule = false;
    </script>
    <script nomodule>
      // nomodule attribute supported
      nomodule = true;
    </script>
    <script>
      if (!nomodule) {
        document.write('<script src="{{ prefix | safe }}/_nicegui/{{version}}/static/es-module-shims.js"><' + '/script>');
    }
    </script>
```

**Comment by falkoschindler** [comment]
@alexc155 It feels like your approach has a logical error: `nomodule` will only be true if the browser understands the "nomodule" attribute _and_ if it doesn't support modules. This leaves two scenarios for `nomodule = false` that are indistinguishable: a) The browser doesn't understand the "nomodule" attribute and, thus, needs shims to play it save. Or b) the browser does understand the "nomodule" attribute but also understands modules and, thus, doesn't need shims.

I decided to remove the "nomodule" altogether in PR #5067. It doesn't cost much and seems to be much more robust than sniffing browser capabilities.

**Comment by alexc155** [comment]
Yes, there appears to be a discrepancy in Mac Monterey between the handling of `nomodule` attribute and whether it actually supports modules!

So Monterey needs the shim library in native mode, but ignores it because it thinks it supports `nomodule`

I think your change to remove `nomodule` altogether is good - the shim code should work with or without system module support

**Comment by alexc155** [comment]
(Also, thank you so much for such a fantastic library - nicegui is a great product!)
